iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -0,0 +1,309 @@
+  };
+  DispatchLinalgOnTensorsPass(const DispatchLinalgOnTensorsPass &pass) {}
+  void runOnOperation() override;
+
+ private:
+  ListOption<int64_t> tileSizes{
+      *this, "tile-sizes", llvm::cl::desc("Set tile sizes to use"),
+      llvm::cl::ZeroOrMore, llvm::cl::MiscFlags::CommaSeparated};
+};
+
+static Operation *buildFlowWorkgroupDispatchOp(OpBuilder &b,
+                                               linalg::LinalgOp root,
+                                               linalg::LinalgOp &clonedRoot) {
+  Location loc = root->getLoc();
+  // TODO(nicolasvasilache): for now this is a 3-D grid of 1's.
+  // In the future make constant and bind late into backend-specific values.
+  SmallVector<Value, 1> count(3, b.create<ConstantIndexOp>(loc, 1));
+
+  SmallVector<Value, 4> dispatchOperands;
+  dispatchOperands.reserve(root->getNumOperands());
+  dispatchOperands.append(root.getInputs().begin(), root.getInputs().end());
+  for (OpOperand &opOperand : root.getOutputOpOperands()) {
+    if (root.isInitTensor(&opOperand)) {
+      dispatchOperands.push_back(opOperand.get());
+    } else {
+      dispatchOperands.push_back(
+          b.create<Shape::GetRankedShapeOp>(loc, opOperand.get()));
+    }
+  }
+  auto otherOperands = root.getAssumedNonShapedOperands();
+  dispatchOperands.append(otherOperands.begin(), otherOperands.end());
+  auto dispatchOp = b.create<IREE::Flow::DispatchWorkgroupsOp>(
+      loc, count, root->getResultTypes(), dispatchOperands);
+  Region &region = dispatchOp.body();
+  Block *block = &region.front();
+  {
+    OpBuilder::InsertionGuard g(b);
+    b.setInsertionPointToStart(block);
+    clonedRoot = cast<linalg::LinalgOp>(b.clone(*root.getOperation()));
+    // Note: DispatchOutputStoreOp is an abstraction jump that consumes the SSA
+    // value produced by `clonedRoot` but it does not comply with the semantics
+    // of DispatchWorkgroupsOp which explicitly states:
+    // "behavior is undefined if multiple workgroups store to the same regions
+    // of the output tensors".
+    // Similarly to sequentialized SPMD loops, the semantics is valid assuming a
+    // sequential ordering of execution.
+    // After destructive update rewrites, the abstraction gap disappears.
+    for (auto it : llvm::zip(clonedRoot->getResults(),
+                             dispatchOp.body().getArguments().take_back(
+                                 clonedRoot->getNumResults()))) {
+      b.create<IREE::Flow::DispatchOutputStoreOp>(loc, std::get<0>(it),
+                                                  std::get<1>(it), llvm::None,
+                                                  llvm::None, llvm::None);
+    }
+    // TODO(nicolasvasilache): return `clonedRoot->getResults()` once we have
+    // shape operands and we drop tie_shape.
+    b.create<IREE::Flow::ReturnOp>(loc);
+  }
+  return dispatchOp;
+}
+
+// Rewrite pattern to ensure only ops with tensor semantics are tiled.
+struct TileAndDistributeOnTensorsPattern
```

Comment:
I am fine with landing this, but this will need to be seriously adapted if we used fusion. We will have to pull in multiple ops before tile + fuse and then tile + fuse.
My initial assumption here was we would just tile + distribute first, then create the dispatch region using the outer most loop.... This still be a phase ordering issue.  Here we would need to
1) Know why ops can be tile+fused
2) Pull them within dispatch region
3) Tile+fuse

Instead it would be better to
1) Tile + fuse
2) Pull each outer loop body into dispatch region.

Reply:
Let's start with this and iterate on a concrete use case if you don't mind.
I went with this implementation because it is significantly more elegant and easier to implement.
We can make it more complex to support relaxing the usage but I didn't want to commit to generality and complexity before I needed to.

I am not sure we have to sequence things like 1 + 2 + 3 though, there may be simpler ways like:
1. move op to dispatch region
2. tile op inside the dispatch region
3. iteratively pull producers by moving them into dispatch region and then fusing.

TBD together in the future.